airflow/sentry.py

```diff
@@ -0,0 +1,196 @@
+
+
+@provide_session
+def get_task_instances(dag_id, task_ids, execution_date, session=None):
+    """
+    Retrieves task instance based on dag_id and execution_date.
+    """
+    from airflow.models.taskinstance import TaskInstance  # Avoid circular import
+
+    if session is None or not task_ids:
+        return []
+
+    TI = TaskInstance
+    return (
+        session.query(TI)
+        .filter(
+            TI.dag_id == dag_id,
+            TI.task_id.in_(task_ids),
+            TI.execution_date == execution_date,
+            or_(TI.state == State.SUCCESS, TI.state == State.FAILED),
+        )
+        .all()
+    )
+
+
+class ConfiguredSentry:
+    """
+    Configure Sentry SDK.
+    """
+
+    SCOPE_TAGS = frozenset(("task_id", "dag_id", "execution_date", "operator"))
+    SCOPE_CRUMBS = frozenset(
+        ("dag_id", "task_id", "execution_date", "state", "operator", "duration")
+    )
+
+    def __init__(self):
+        """
+        Initialize the Sentry SDK.
+        """
+        ignore_logger("airflow.task")
+        ignore_logger("airflow.jobs.backfill_job.BackfillJob")
+        executor_name = conf.get("core", "EXECUTOR")
+
+        sentry_flask = FlaskIntegration()
+
+        # LoggingIntegration is set by default.
+        integrations = [sentry_flask]
+
+        if executor_name == "CeleryExecutor":
+            from sentry_sdk.integrations.celery import CeleryIntegration
+            from sentry_sdk.integrations.tornado import TornadoIntegration
+
+            sentry_celery = CeleryIntegration()
+            sentry_tornado = TornadoIntegration()
+            integrations += [sentry_celery, sentry_tornado]
+
+        dsn = conf.get("sentry", "sentry_dsn")
+        if dsn:
+            init(dsn=dsn, integrations=integrations)
+        else:
+            # Setting up Sentry using environment variables.
+            log.debug("Defaulting to SENTRY_DSN in environment.")
+            init(integrations=integrations)
+
+    def add_tagging(self, task_instance):
+        """
+        Function to add tagging for a task_instance.
+        """
+        task = task_instance.task
+
+        with configure_scope() as scope:
+            for tag_name in self.SCOPE_TAGS:
+                attribute = getattr(task_instance, tag_name)
+                if tag_name == "operator":
+                    attribute = task.__class__.__name__
+                scope.set_tag(tag_name, attribute)
+
+    @provide_session
+    def add_breadcrumbs(self, task_instance, session=None):
+        """
+        Function to add breadcrumbs inside of a task_instance.
+        """
+        if session is None:
+            return
+        dag_id = task_instance.dag_id
+        execution_date = task_instance.execution_date
+        task = task_instance.task
+        dag = task.dag
+        task_ids = dag.task_ids
+        task_instances = get_task_instances(
+            dag_id, task_ids, execution_date, session=session
+        )
+
+        for ti in task_instances:
```

Comment:
This function isn't clear to me what it's doing (I'm possibly not that used to Sentry).

Why does it take a single task_instance, but then walk over all the task instances of the dag?

Reply:
It's also not clear to me why we want all the completed TIs for a task. 

Can you show where this shows up in Sentry?

Do we instead here want to look at just the upstrream tasks of this task? Do we actually need more than the DAG_id in the breadcrumb?

Reply:
Okay, that makes some sense. I feel a screen shot or two in the doc might help this.

Reply:
![Breadcrumbs](https://lh4.googleusercontent.com/kcZojgIjMrn1TleiUMkdkkgEJ40l_tvlPlQx81alQu-qEzHYMk9yG1we_ZAlh0tqIR9WD1H7n0HnVX6XoZot=w2688-h1596-rw)
Here is an example of the breadcrumbs in action. They give extra details on the duration of previous and concurrent task instances.

Reply:
@tiopi I can't see that image (anymore?)